### PR TITLE
chore: Bump Koin 1.0.2 to version 2.0.0-GA4

### PR DESCRIPTION
**Pull request by anhanh11001**
Details:
- Fix imports
- Remove Scope
- Fix build by improving database schema

@iamareebjamal There is a part of the dependency test where there is no dryRun() function anymore. I don't know how to fix that so, for now, I'm commenting it out

### CODE REVIEW

**Comment by iamareebjamal** [comment]
Read the migration guide. There's a replacement for that

**Comment by anhanh11001** [comment]
My build is not successful because in the dependency test, `    viewModel { GeoLocationViewModel(get()) }` is not instantiated successfully, due to NullPointerException. I have updated the latest version of this issue @liveHarshit @iamareebjamal. Please give some advices, I couldn't figure out what went wrong.
```
org.koin.core.error.InstanceCreationException: Could not create instance for [type:Factory,primary_type:'org.fossasia.openevent.general.search.GeoLocationViewModel']
	at org.koin.core.instance.DefinitionInstance.create(DefinitionInstance.kt:61)
	at org.koin.core.instance.FactoryDefinitionInstance.get(FactoryDefinitionInstance.kt:37)
	at org.koin.core.definition.BeanDefinition.resolveInstance(BeanDefinition.kt:71)
	at org.koin.core.scope.Scope.resolveInstance(Scope.kt:140)
	at org.koin.core.scope.Scope.get(Scope.kt:130)
	at org.koin.core.Koin.get(Koin.kt:107)
	at org.koin.test.check.CheckModulesKt.checkMainDefinitions(CheckModules.kt:60)
	at org.koin.test.check.CheckModulesKt.checkModules(CheckModules.kt:38)
	at org.koin.test.check.CheckModulesKt.checkModules(CheckModules.kt:25)
	at org.koin.test.check.CheckModulesKt.checkModules$default(CheckModules.kt:25)
	at org.fossasia.openevent.general.DependencyTest.testDependencies(DependencyTest.kt:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:106)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:155)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:137)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at androidx.arch.core.executor.DefaultTaskExecutor.isMainThread(DefaultTaskExecutor.java:77)
	at androidx.arch.core.executor.ArchTaskExecutor.isMainThread(ArchTaskExecutor.java:116)
	at androidx.lifecycle.LiveData.assertMainThread(LiveData.java:460)
	at androidx.lifecycle.LiveData.setValue(LiveData.java:304)
	at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
	at org.fossasia.openevent.general.search.GeoLocationViewModel.<init>(GeoLocationViewModel.kt:19)
	at org.fossasia.openevent.general.di.ModulesKt$viewModelModule$1$23.invoke(Modules.kt:223)
	at org.fossasia.openevent.general.di.ModulesKt$viewModelModule$1$23.invoke(Modules.kt)
	at org.koin.core.instance.DefinitionInstance.create(DefinitionInstance.kt:54)
	... 57 more
```

**Comment by liveHarshit** [comment]
> My build is not successful because in the dependency test

Build failing on Android studio or Travis?

**Comment by iamareebjamal** [comment]
There's a null pointer exception happening in the constructor of the ViewModel GeoLocationViewModel.kt:19

File: app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt [hunk 1]
```diff
@@ -9,11 +9,10 @@ import androidx.room.Query
...
 interface SponsorWithEventDao {
     @Query("""
-        SELECT * FROM sponsor
-        INNER JOIN sponsorwithevent ON
-        sponsor_id = sponsorwithevent.sponsor_id
-        WHERE sponsorwithevent.event_id = :eventID
-        """)
+        SELECT * FROM sponsor s
+        WHERE EXISTS (SELECT * FROM sponsorwithevent se
+        WHERE s.id = se.sponsor_id AND se.event_id = :eventID)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt]: This is suboptimal. Please revert db changes from this PR. It should be focused on Koin only

File: app/src/test/java/org/fossasia/openevent/general/DependencyTest.kt [hunk 2]
```diff
@@ -6,20 +6,23 @@ import org.fossasia.openevent.general.di.commonModule
...
-        // start Koin
-        startKoin(listOf(commonModule, apiModule, viewModelModule,
-            networkModule, databaseModule)) with mock(Application::class.java)
-        // dry run of given module list
-        dryRun()
+
+        startKoin {
+            androidLogger()
+            androidContext(mock(Application::class.java))
+            modules(commonModule, apiModule, viewModelModule, networkModule, databaseModule, flavorSpecificModule)
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/general/DependencyTest.kt]: See previous list, don't include flavorSpecificModule

File: app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt (hunk 1 above)
Comment (anhanh11001) [reply app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt]: it's also a problem in the travis build where the query used before contains excessive columns

File: app/src/test/java/org/fossasia/openevent/general/DependencyTest.kt (hunk 2 above)
Comment (anhanh11001) [reply app/src/test/java/org/fossasia/openevent/general/DependencyTest.kt]: yes I will

File: app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt (hunk 1 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt]: Please paste the error here

Comment (anhanh11001) [reply app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt]:
```
w: /home/travis/build/fossasia/open-event-android/app/build/tmp/kapt3/stubs/fdroidDebug/org/fossasia/openevent/general/speakers/SpeakerWithEventDao.java:11: warning: The query returns some columns [event_id, speaker_id] which are not used by org.fossasia.openevent.general.speakers.Speaker. You can use @ColumnInfo annotation on the fields to specify the mapping.  You can suppress this warning by annotating the method with @SuppressWarnings(RoomWarnings.CURSOR_MISMATCH). Columns returned by the query: id, name, email, photoUrl, shortBiography, longBiography, speakingExperience, position, mobile, location, country, city, organisation, gender, website, twitter, facebook, linkedin, github, isFeatured, event_id, speaker_id. Fields in org.fossasia.openevent.general.speakers.Speaker: id, name, email, photoUrl, shortBiography, longBiography, speakingExperience, position, mobile, location, country, city, organisation, gender, website, twitter, facebook, linkedin, github, isFeatured.
    public abstract androidx.lifecycle.LiveData<java.util.List<org.fossasia.openevent.general.speakers.Speaker>> getSpeakerWithEventId(long eventID);
```

Comment (anhanh11001) [reply app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt]:
```
w: /home/travis/build/fossasia/open-event-android/app/build/tmp/kapt3/stubs/fdroidDebug/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.java:11: warning: The query returns some columns [event_id, sponsor_id] which are not used by org.fossasia.openevent.general.sponsor.Sponsor. You can use @ColumnInfo annotation on the fields to specify the mapping.  You can suppress this warning by annotating the method with @SuppressWarnings(RoomWarnings.CURSOR_MISMATCH). Columns returned by the query: id, name, description, url, logoUrl, level, type, event_id, sponsor_id. Fields in org.fossasia.openevent.general.sponsor.Sponsor: id, name, description, url, logoUrl, level, type.
    public abstract androidx.lifecycle.LiveData<java.util.List<org.fossasia.openevent.general.sponsor.Sponsor>> getSponsorWithEventId(long eventID);
```

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt]: These are just warnings. Will tackle them later and in a separate issue

Comment (anhanh11001) [reply app/src/main/java/org/fossasia/openevent/general/sponsor/SponsorWithEventDao.kt]: I'll make a new issue now

**Comment by anhanh11001** [comment]
@iamareebjamal updated. There is a part where I don't quite understand. So previously, the NullPointerException was in this part in the init block. So I guess mutableVisibility was null somehow. I fixed this by making default value in the constructor of LiveData object:
`    private val mutableVisibility = MutableLiveData<Boolean>() ->     private val mutableVisibility = MutableLiveData<Boolean>(false)` then it works. Can you explain this or give me some keywords to read more about this behaviour?

```
class GeoLocationViewModel(locationService: LocationService) : ViewModel() {
    private val mutableLocation = MutableLiveData<String>()
    val location: LiveData<String> = mutableLocation
    private val mutableVisibility = MutableLiveData<Boolean>()
    val currentLocationVisibility: LiveData<Boolean> = mutableVisibility
    private val mutableOpenLocationSettings = MutableLiveData<Boolean>()
    val openLocationSettings: LiveData<Boolean> = mutableOpenLocationSettings
    private val mutableErrorMessage = SingleLiveEvent<String>()
    val errorMessage: LiveData<String> = mutableErrorMessage

    init {
        mutableVisibility.value = false // NullPointerException at this line
    }
```

**Comment by iamareebjamal** [comment]
It was simply giving error because you can't use GeoLocationViewModel in tests because it depends on Android classes

**Pull request merged**